compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -467,175 +468,235 @@ private void checkXMLNamespacePrefixes(List<BLangXMLElementFilter> filters) {
-                    setLiteralValueForFiniteType(literalExpr, valueType);
-                    return valueType;
-                } else if (literalAssignableToFiniteType(literalExpr, finiteType, TypeTags.SIGNED8_INT)) {
-                    BType valueType = setLiteralValueAndGetType(literalExpr, symTable.signed8IntType);
-                    setLiteralValueForFiniteType(literalExpr, valueType);
-                    return valueType;
-                } else if (literalAssignableToFiniteType(literalExpr, finiteType, TypeTags.UNSIGNED32_INT)) {
-                    BType valueType = setLiteralValueAndGetType(literalExpr, symTable.unsigned32IntType);
-                    setLiteralValueForFiniteType(literalExpr, valueType);
-                    return valueType;
-                } else if (literalAssignableToFiniteType(literalExpr, finiteType, TypeTags.UNSIGNED16_INT)) {
-                    BType valueType = setLiteralValueAndGetType(literalExpr, symTable.unsigned16IntType);
-                    setLiteralValueForFiniteType(literalExpr, valueType);
-                    return valueType;
-                } else if (literalAssignableToFiniteType(literalExpr, finiteType, TypeTags.UNSIGNED8_INT)) {
-                    BType valueType = setLiteralValueAndGetType(literalExpr, symTable.unsigned8IntType);
-                    setLiteralValueForFiniteType(literalExpr, valueType);
-                    return valueType;
-                }
-            } else if (expType.tag == TypeTags.UNION) {
-                Set<BType> memberTypes = ((BUnionType) expType).getMemberTypes();
-                BType intSubType = null;
-                boolean intOrIntCompatibleTypeFound = false;
-                for (BType memType : memberTypes) {
-                    if ((memType.tag != TypeTags.INT && TypeTags.isIntegerTypeTag(memType.tag)) ||
-                            memType.tag == TypeTags.BYTE) {
-                        intSubType = memType;
-                    } else if (memType.tag == TypeTags.INT || memType.tag == TypeTags.JSON ||
-                            memType.tag == TypeTags.ANYDATA || memType.tag == TypeTags.ANY) {
-                        intOrIntCompatibleTypeFound = true;
-                    }
-                }
-                if (intOrIntCompatibleTypeFound) {
-                    return setLiteralValueAndGetType(literalExpr, symTable.intType);
-                }
-                if (intSubType != null) {
-                    return setLiteralValueAndGetType(literalExpr, intSubType);
+    private int getPreferredMemberTypeTag(BFiniteType finiteType) {
+        for (BLangExpression valueExpr : finiteType.getValueSpace()) {
+            int typeTag = valueExpr.getBType().tag;
+            if (typeTag > TypeTags.DECIMAL) {
+                continue;
+            }
+            for (int i = TypeTags.INT; i <= TypeTags.DECIMAL; i++) {
+                if (typeTag == i) {
+                    return i;
                 }
```

Comment:
Can't we just use an if check or a switch statement here instead? Would help with readability also, IMO.